src/libraries/Common/tests/System/Net/Http/Http3LoopbackConnection.cs

```diff
@@ -44,21 +50,26 @@ public Http3LoopbackConnection(QuicConnection connection)
 
         public override void Dispose()
         {
+            // Close any remaining request streams (but NOT control streams, as these should not be closed while the connection is open)
             foreach (Http3LoopbackStream stream in _openStreams.Values)
             {
                 stream.Dispose();
             }
 
-            if (!_closed)
-            {
-            //    CloseAsync(H3_INTERNAL_ERROR).GetAwaiter().GetResult();
-            }
+            // Dispose the connection
+            // If we already waited for graceful shutdown from the client, then the connection is already closed and this will simply release the handle.
+            // If not, then this will silently abort the connection.
+            _connection.Dispose();
```

Comment:
Will this cause us to drop connections prematurely again? I know that the close/dispose were commented out, because the server task would sometimes finish before the client. It would close the connection from the server side and then the client response reading would throw due to a closed stream.

Reply:
I don't think so. 

The original problem was with code that would use CreateClientAndServer. The problem here was that we would spawn a task for the client and another for the server, then wait for them both. And the server task would dispose the server connection when it completed. This could result in losing data. The fix was to have the server wait for graceful shutdown in these cases.

Most code that doesn't use CreateClientAndServer should not have this problem.

Note though that if code does have this problem, then it already has it because even though we didn't dispose the connection here previously, it was still collectible at this point, meaning that we would likely see occasional failures here. So actively disposing the connection here is probably better since it will be more likely to fail consistently.

Reply:
I think that we might revisit some of the H/3 tests and see if this might be a problem there. For example: 
https://github.com/dotnet/runtime/blob/c8dc49e6e7a654cc730a18c209513561f560c1ae/src/libraries/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.Http3.cs#L35

It's disabled now so we wouldn't see any problem, but I think it might suffer from this problem. I assume that adding `WaitForClientDisconnectAsync` add the end of the server task should mitigate this, am I right?

Reply:
I agree, there could be a problem here, unfortunately.

> I assume that adding WaitForClientDisconnectAsync at the end of the server task should mitigate this, am I right?

Yes, I believe that would handle it. That said, I think we should revisit how this works generally, as it is slightly different for every version of HTTP, which seems bad. That's not something to tackle for 6.0 though.

I think what we should do here is just revert the change here that disposes the connection, to minimize behavioral change in the loopback server. I'll update and post a new version of the PR.